src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -90,8 +90,17 @@ ErrorMessageOr<uint64_t> ExecuteOrDie(pid_t pid, uint64_t address_code, uint64_t
   // The calling convention also asks for rsp being a multiple of 16 so we additionally round down
   // to the next multiople of 16.
   const uint64_t old_rsp = original_registers.GetGeneralPurposeRegisters()->x86_64.rsp;
-  const uint64_t aligned_rsp_below_red_zone = (old_rsp - 128) / 16 * 16;
+  constexpr int kSizeRedZone = 128;
+  constexpr int kStackAlignment = 16;
+  const uint64_t aligned_rsp_below_red_zone =
+      ((old_rsp - kSizeRedZone) / kStackAlignment) * kStackAlignment;
```

Comment:
I know this actually belongs to the previous CL, but this can be tested right? Have puppet allocate something prime-number aligned on the stack and check that it doesn't crash?

Reply:
and I still think & ~(alignment - 1) looks better.

Reply:
maybe even introduce constexpr helper function that aligns arbitrary value to a power of 2. (and static_assert for power_of_two)

Reply:
> and I still think & ~(alignment - 1) looks better

I actually find dividing and multiplying easier to read (it's encountered much more often as it works for any alignment).

Reply:
> I actually find dividing and multiplying easier to read

Same here but I have to admit that I saw 
`and $0xfffffff0,%esp`
in handwritten asssembler code and also emitted from the compiler in the last few days.

Reply:
note that this is highly unusual to align something to not a power of two value. So it would be nice to have a check in align() function.

Reply:
> , but this can be tested right?

For a meaningful test I would need something that crashes (without the enforced alignment) first.
I spend some time to produces a crash due to this but - surprisingly - didn't succeed. Even in the trivial example the 16 byte alignment is violated very often; I see rsp ending in "8" before the call and the function is happily executed (with the aligning commented out).
My understanding is that there are two reasons to enforce the alignment: performance penilties for unaligned loads and SIMD instructions that crash on unaligned access. It is surprisingly hard to get the later from within C++ - even with intrinsics the compiler tends to shuffle things around such that nothing bad happens. 
I then wrote inline assembly to cause a misaligned access on the stack:

  ` __asm__ __volatile__("movaps -7(%%rsp),%%xmm1\n\t" : : : "%rsp");`

This this does not crash either although the intel manual clearly states that "a general-protection exception (#GP) will be generated". 

I'll spend a bit more time today looking into this but currently I don't have a test for this.